### ISSUE TITLE
False report of u2u1023

### ISSUE THREAD

**Issue by wmjordan** [u2uconsult/codeanalyzers#33]
The following code snippet triggers u2u1023 (_The variable 's' is initialized but isn't read before it's overwritten, consider removing the initializer_), which is incorrect.

````C#
string Do() {
  string s = GetString();
  fixed (char* p = s) {
  // do something with p
  }
  return s;
}
````

**Comment by KrisVandermotten** [comment]
This may be caused by a bug in Roslyn, see https://github.com/dotnet/roslyn/issues/60402

We'll try to find a workaround.

**Comment by wmjordan** [comment]
I think Roslyn's current implementation is partially correct.

Once we obtain a reference to the string, we can not only read but also modify its content, so it can be _WrittenInside_ the fixed statement. Actually the above snippet was stripped from a method that used unsafe pointers to modify the string.

**Comment by KrisVandermotten** [comment]
> I think Roslyn's current implementation is partially correct.

In the mean time, the Roslyn team confirmed the bug. (Also, I don't know what "partially correct" means.)

> Once we obtain a reference to the string, we can not only read but also modify its content, so it can be WrittenInside the fixed statement. Actually the above snippet was stripped from a method that used unsafe pointers to modify the string.

Modifying the content of an object is not the same thing as modifying a variable that contains a reference to an object.

For example, assume there is a class Person and a variable p of that type:
```cs
class Person
{
    public string name;
}

Person p;
```

Then this is writing p:
```cs
p = new Person();
```
and this is not:
```cs
p.name = "Kris";
```
Actually, the latter is reading `p` in order to dereference it, i.e. to find the location of `p.name`, and then it writes `p.name`.

**Comment by KrisVandermotten** [comment]
> Actually the above snippet was stripped from a method that used unsafe pointers to modify the string.

You really should not be doing that. See https://github.com/dotnet/runtime/blob/main/docs/design/features/StringDeduplication.md for at least one reason not to do that:

> Strings are supposed to be immutable. However in unsafe code you can change the string content after it’s created. Changing string content already asking for trouble without deduping – you could be changing the interned copy which means you are modifying someone else’s string which could cause completely unpredictable results for them.

If you want to write to the string while creating it, and you're on .NET Core 2.1 or later, use [string.Create](https://docs.microsoft.com/en-us/dotnet/api/system.string.create?view=net-6.0#system-string-create-1(system-int32-0-system-buffers-spanaction((system-char-0)))).

**Comment by wmjordan** [comment]
Thank you for your kind advice. I am aware of that.

For my scenarios, "string mutation" is only used while they are created. And I don't mutate strings not created from my part.

One application is to convert `DateTime` value to an ad hoc format, for instance "yyyy-MM-dd", in a fast way. 
The implementation works like that:
1. allocate a string with 10 chars.
2. place year, month, day and two `-`s into there.
    1. the yyyy can be divided into two character pairs, so to the month and day part
    2. so, we can pre-generate and cache them, then paste those character pairs onto the string
3. return the string

I still have to target .NET Framework 4, thus `String.Create` can not be used either.

**Comment by KrisVandermotten** [comment]
This is fixed in version 3.8.1.

**Comment by wmjordan** [comment]
Thank you for taking time to fix this issue.
However, if more than one `fixed` statements are involved, this issue can still occur (on the `t` in the following snippet).

````C#
unsafe string Do() {
  string s = GetString(), t = GetString2();
  fixed (char* p = s)
  fixed (char* p2 = t) {
  // do something with p
  }
  return t;
}
````

**Comment by KrisVandermotten** [comment]
> However, if more than one fixed statements are involved, this issue can still occur (on the t in the following snippet).

That is correct. Please bear in mind that we're trying to work around a Roslyn bug here, while waiting for a proper fix from Microsoft.

Version 3.8.2 will fix your specific example, or you can work around the problem by writing it as follows:

```cs
unsafe string Do()
{
  string s = GetString(), t = GetString2();
  fixed (char* p = s, p2 = t)
  {
    // do something with p
  }
  return t;
}
```

In other cases, you might have to 
```cs
#pragma warning disable U2U1023
```
while we wait for Microsoft to fix the bug.

**Comment by wmjordan** [comment]
Thank you for your explanation.
It is not a severe issue for me, since I know what i am doing.
Let's wait for MS or the community to fix it.
